### PR TITLE
*: refactor all C bindings

### PR DESCRIPTION
**Pull request by BusyJay**
- Wrapper types are removed so we can use the raw pointers directly, such
  as DB* instead of crocksdb_db*
- Convert all global variables to function access. Otherwise rustc may not
  do correct job to keep necessary symbols and cause link errors
- Introduce a byte receiver to reduce one memory allocation between ABI
- Divide crocksdb_options_t into DBOptions and CFOptions
- Use predefine.rs to define custom types so we can generate correct regex
  instead of hard code
- Update rocksdb to match the same version in tikv/rust-rocksdb

To see how the APIs are used, please take a look at dev branch.

### CODE REVIEW

File: tirocks-sys/build.rs [hunk 1]
```diff
@@ -22,6 +22,30 @@ fn bindgen_rocksdb(file_path: &Path) {
     if env::var("CARGO_CFG_TARGET_OS").map_or(false, |s| s == "windows") {
         builder = builder.clang_arg("-D _WIN32_WINNT=0x600");
     }
+    let pre_defined: Vec<_> = std::fs::read_to_string("src/pre_defined.rs")
+        .unwrap()
+        .lines()
+        .map(|l| {
+            let mut new_string = String::new();
+            let mut visited = false;
+            for part in (l[17..]).split_whitespace().next().unwrap().split('_') {
```
Comment (tabokie) [inline tirocks-sys/build.rs]:
```suggestion
            for part in (l["pub enum rocksdb_".len()..]).split_whitespace().next().unwrap().split('_') {
```

File: tirocks-sys/src/lib.rs [hunk 2]
```diff
@@ -12,24 +12,43 @@ extern crate bzip2_sys;
...
 
-use std::{ffi::CStr, str::Utf8Error};
+use std::{ffi::CStr, mem, str::Utf8Error};
 
 pub use bindings::*;
 
+/// Check if it's safe to convert transmute from `&[u8]` to `rocksdb_Slice`.
+///
+/// Following code should be optimized away in release build.
+pub fn rocks_slice_same_as_rust() -> bool {
```
Comment (tabokie) [inline tirocks-sys/src/lib.rs:28]: This is no used?

File: tirocks-sys/src/lib.rs [hunk 3]
```diff
@@ -150,6 +169,13 @@ impl rocksdb_Status {
     }
 }
 
+impl PartialEq for rocksdb_Status {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        self.code_ == other.code_
```
Comment (tabokie) [inline tirocks-sys/src/lib.rs:175]: add `subcode_` and `sev_`

File: tirocks-sys/src/lib.rs (hunk 2 above)
Comment (BusyJay) [reply tirocks-sys/src/lib.rs:28]: Will be used later.

File: tirocks-sys/src/lib.rs (hunk 3 above)
Comment (BusyJay) [reply tirocks-sys/src/lib.rs:175]: The native implementation only checks for code.  https://github.com/tikv/rocksdb/blob/6.4.tikv/include/rocksdb/status.h#L371

**Review by tabokie — APPROVED** [review]

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**